### DISCUSSION TITLE
AI Deck and Loco Positioning Deck have GPIO Conflicts

### DISCUSSION THREAD

**Discussion by 787-10** [bitcraze/discussions#854] [Q&A] [answered]
Hello, I am currently experiencing some issues with having the AI deck and Loco Positioning Deck working together on the Crazyflie 2.1.

My entire setup will consist of three decks -- the AI Deck, the Loco Positioning Deck, and the Flow Deck. Currently, the AI deck and flow deck work together perfectly, and the flow deck also works perfectly when coupled to the loco positioning deck. However, no decks seem to be initialized if the AI deck and loco positioning deck are ever connected together; in the cfclient, I get the following console message:

`DECK_INFO: ERROR: Driver Gpio usage conflicts with a previously enumerated deck driver. No decks will be initialized!`

It seems like from this error the AI deck might be using the same GPIO pins that the Loco positioning deck is already using. What can be done to configure the AI deck and/or loco positioning deck such that they no longer have a GPIO pin conflict? Will it be possible to configure all of the GPIO pins such that the flow deck, AI deck, and loco positioning deck are all working together, and what would be an optimal GPIO configuration for my intended setup?

Thank you very much.

**Comment by 787-10** [comment]
Hello, just an update on the issue. I have referenced the deck compatibility matrix as well as [this thread here](https://forum.bitcraze.io/viewtopic.php?t=5188) and if my understanding is correct, I believe that GPIO 1 seems to be used by both the Loco Positioning Deck and the AI Deck. Since the flow deck seems to be using GPIO 3 and that the AI Deck is using GPIO1 and GPIO4, I think we may need to do the patching on the loco deck from GPIO1 to GPIO2.

Attached please find a picture of the underside of the drone with the bottom of the Loco Positioning Deck facing the camera. I notice that there are 4 pins labeled GS1 through GS4 - do these correspond to GPIO1-4? And for each of these labeled pins, I see two rectangular metallic areas that seem to be separated. How should I build a solder bridge (or use a breadboard) to switch the pin that the loco deck is using from GPIO 1 to GPIO 2?

<img width="605" alt="Screenshot 2023-08-02 at 8 53 01 PM" src="https://github.com/bitcraze/crazyflie-firmware/assets/60076597/9d46ceb2-0f83-4628-98d8-fe2c5abde5fa">

Additionally, for the firmware update, I am changing the following pieces of code in src/deck/drivers/src/locodeck.c:
`#define CS_PIN DECK_GPIO_IO1` to `#define CS_PIN DECK_GPIO_IO2`
Here I noticed that GPIO2 is labeled as an IRQ pin and that 3 and 4 are labeled as RESET pins -- does this mean that I can switch to using GPIO2 instead of GPIO1 as the CS pin?

Finally, would I have to change the following to reflect the above changes as well?
```
#ifdef CONFIG_DECK_LOCODECK_USE_ALT_PINS
  #ifndef CONFIG_LOCODECK_ALT_PIN_RESET
  .usedGpio = DECK_USING_IO_1 | DECK_USING_IO_2 | DECK_USING_IO_3,
  #else
  .usedGpio = DECK_USING_IO_1 | DECK_USING_IO_2 | DECK_USING_IO_4,
  #endif
#else
   // (PC10/PC11 is UART1 TX/RX)
  .usedGpio = DECK_USING_IO_1 | DECK_USING_PC10 | DECK_USING_PC11,
#endif
```
to the following:
```
#ifdef CONFIG_DECK_LOCODECK_USE_ALT_PINS
  #ifndef CONFIG_LOCODECK_ALT_PIN_RESET
  .usedGpio = DECK_USING_IO_1 | DECK_USING_IO_2 | DECK_USING_IO_3,
  #else
  .usedGpio = DECK_USING_IO_1 | DECK_USING_IO_2 | DECK_USING_IO_4,
  #endif
#else
   // (PC10/PC11 is UART1 TX/RX)
  .usedGpio = DECK_USING_IO_2 | DECK_USING_PC10 | DECK_USING_PC11,
#endif
```

Any tips on this would be greatly appreciated, thank you very much in advance!

**Comment by 787-10** [comment]
Finally, I am curious to know if the orientation of the decks has to be changed after I patch the loco positioning deck. Currently, I have the orientation set as the following:

Below the main body of the drone, I have the flow deck, and immediately above it, I have the Loco Positioning Deck. Then, the AI deck is above the main body of the drone (above the battery).

I was curious to see if there may be a need to change the orientation to have only the flow deck be on the bottom half and for the loco positioning deck, then the AI deck to be on the top half. If we're assuming that the flow of electricity is from the board upwards, then it would reach the loco positioning deck first and eventually the AI deck. On the loco deck, I would build a solder bridge connecting GS1 and GS2, and I guess GS 1 will be used for the AI deck.

**Comment by tobbeanton** [comment]
I'm not sure yet but I think it might be possible to get them running together without any hardware modifications. The BOOT signal for the AI-deck is only used att boot so it might be possible to get them working together with just some firmware modifications. I will try to find some time this week to get it to work.

**Reply by tobbeanton** [reply, marked as answer]
It was a lot easier then I thought. Just remove the DECK_USING_IO_1 in aideck.c
```
static const DeckDriver aideck_deck = {
    .vid = 0xBC,
    .pid = 0x12,
    .name = "bcAI",

    .usedGpio = DECK_USING_IO_4,
    .usedPeriph = DECK_USING_UART2,

    .memoryDef = &espMemoryDef,
    .memoryDefSecondary = &gap8MemoryDef,

    .init = aideckInit,
    .test = aideckTest,
};
```

**Reply by tobbeanton** [reply]
It is going to be heavy so I think you will need the "thrust upgrade kit" to get it to fly reliably. Perhaps a Tattu 350mAh battery as well.

**Comment by 787-10** [comment]
Hello @tobbeanton, thank you very much for your reply. We were able to resolve the GPIO related errors and now the Crazyflie is detecting all three decks again -- in particular, the AI and loco can be connected together and I am able to use wifi-related tasks on the AI deck like streaming video, etc.

However, when I connect the flow deck, another problem seems to arise. The DW1000 on the loco is failing to initialize, and tracking this down further seems to indicate that the data is not being properly read from the DW1000's SPI pin. This is not occurring when I have the AI deck and loco all together, but only when all three decks are attached so it seems the flow deck might be interfering with the SPI onboard the loco? The flow deck itself seems to be initializing properly though.

What are some possible solutions around the SPI issue? Is it possible to have both decks transmit at staggered intervals through that pin so the messages do not conflict?

**Reply by tobbeanton** [reply]
Is this without any hardware modifications and just the firmware change I suggested? Tomorrow I will hopefully have time to try and replicate it.

**Reply by 787-10** [reply]
Yes, we have not made any hardware modifications to date.

**Reply by tobbeanton** [reply]
I could not replicate it. For me it works fine. I'm running the latest crazyflie-firmware master with the only change I proposed above. Could you describe in more detail what you are running?

Some takeaways though: 
- It barely takes off and easily gets unstable -> The thrust upgrade kit is a must.
- It might be better to have the Loco deck on top and the AI-deck in the middle since the antenna on the ESP module on the AI-deck is a bit fragile.

**Reply by 787-10** [reply]
Hey Tobias, no worries and thanks for taking the time to try anyways. I realized that there was a hardware issue after taking a Voltmeter to it (we messed with the flow deck so some of the connections were not stable). Those joints were resoldered and I am no longer experiencing the SPI bus failure; I suspect it was either the MISO or MOSI pin that got messed up and wasn't having a solid connection.

Regarding your takeaways:
(1) We noticed that the thrust upgrade kit on your website is currently sold out, so we took down the motors and propellers and are attempting to order them separately. We have a shipment of the motors on the way, but it seems that the size of the pin on the motors is larger than the ones for the stock propulsion kit, so I'm curious to see if it is possible to purchase those rotors separately. I have done some basic flight tests with the current thrust and sometimes it is a bit unstable (especially if the battery is running low), but other times it still is able to complete the task.

(2) This is a good point and we came to the same realization. I have the AI deck and Loco on the top, and below the main STM of the drone is the flow deck. So far we are happy with this configuration and have observed no further issues.
